### ISSUE TITLE
Compiler panics on use of duplicate implementations of a custom trait

### ISSUE THREAD

**Issue by Izodn** [rust-lang/rust#24073]
Compiler panics when compiling code that implements a trait for a generic type and a generic type in a vector.

``` Rust
trait MyTrait {
    fn foo(&self){}
}
trait OtherTrait {
    fn bar(&self){}
}

impl OtherTrait for String {}
impl OtherTrait for Vec<String> {}

impl<T: OtherTrait> MyTrait for T {}
impl<T: OtherTrait> MyTrait for Vec<T> {}

fn main() {
    let test_var: Vec<String> = Vec::<String>::new();
    test_var.foo();
}
```

I expected to see this happen: As in 1.0.0-nightly (3e4be02b8 2015-03-13) (built 2015-03-14), the compiler should catch the conflicting implementations.

Instead, this happened: Compiler panics

## Meta

`rustc --version --verbose`:
rustc 1.0.0-beta (9854143cb 2015-04-02) (built 2015-04-02)
binary: rustc
commit-hash: 9854143cba679834bc4ef932858cd5303f015a0e
commit-date: 2015-04-02
build-date: 2015-04-02
host: x86_64-pc-windows-gnu
release: 1.0.0-beta

Error: internal compiler error: coherence failed to report ambiguity: cannot locate the impl of the trait `MyTrait` for the type `collections::vec::Vec<collections::string::String>`

**Comment by pmarcelll** [comment]
I can't reproduce it with `rustc 1.2.0-nightly (20d23d8e5 2015-06-18)`.

**Comment by alexcrichton** [comment]
Agreed!
